### PR TITLE
[Inductor] Add prologue fusion

### PR DESCRIPTION
**Pull request by michihirohorie**
Hi, I am new in PyTorch, but trying to add prologue fusion in the Inductor-Triton path by following the implementation of epilogue fusion. 

I mainly changed two parts: scheduler and codegen. Current scheduler gives up fusing prologues such as relu followed by mm due to the mismatch of their dependency types. I changed TemplateBuffer so that it also issues MemoryDep when prologue fusion is enabled in config to make the dependency types match.

Regarding codegen, as a first use case, I changed Triton mm kernel so that it can emit prologue code. Kernel is supposed to receive single type of activation function either as its first parameter, second parameter, or both.

Any comments or suggestions are appreciated.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/codegen/cuda/cuda_cpp_scheduling.py [hunk 1]
```diff
@@ -181,7 +181,7 @@ def define_kernel(self, src_code: str, node_schedule) -> str:
         return kernel_name
 
     def codegen_template(
-        self, template_node: BaseSchedulerNode, epilogue_nodes: List[SchedulerNode]
+        self, template_node: BaseSchedulerNode, epilogue_nodes: List[SchedulerNode], isEpilogue=True
```
Comment (jansel) [inline torch/_inductor/codegen/cuda/cuda_cpp_scheduling.py:184]: isEpilogue => is_epilogue

File: torch/_inductor/codegen/common.py [hunk 2]
```diff
@@ -1525,13 +1525,18 @@ def load(name: str, index: sympy.Expr) -> CSEVariable:
...
+
             @staticmethod
             def store(
                 name: str, index: sympy.Expr, value: CSEVariable, mode: StoreMode = None
             ) -> None:
                 self.store_buffer_names.add(name)
                 if mode is None:
-                    self.cse.store_cache[name] = value
+                    if name not in self.cse.store_cache:
+                        self.cse.store_cache[name] = value
```
Comment (jansel) [inline torch/_inductor/codegen/common.py:1539]: Can you explain why this change is needed?

File: torch/_inductor/codegen/common.py [hunk 3]
```diff
@@ -1525,13 +1525,18 @@ def load(name: str, index: sympy.Expr) -> CSEVariable:
                     return store_cache[name]
                 return self.load(name, index)
 
+            @staticmethod
+            def cache_assign(name, value):
```
Comment (jansel) [inline torch/_inductor/codegen/common.py:1529]:
```suggestion
            def set_cse_store_cache(name, value):
```

File: torch/_inductor/codegen/triton.py [hunk 4]
```diff
@@ -1264,6 +1266,8 @@ def __init__(
         self.block_ptr_id = itertools.count()
         # buffer accesses in the kernel
         self.buf_accesses: DefaultDict[str, List[Dep]] = collections.defaultdict(list)
+        self.isEpilogue = True
```
Comment (jansel) [inline torch/_inductor/codegen/triton.py:1269]:
```suggestion
        self.is_epilogue = True
```

File: test/inductor/test_max_autotune.py [hunk 5]
```diff
@@ -805,6 +805,52 @@ def fn(
...
+            mul0 = torch.rand(M0, N).cuda().half()
+            y0 = compiled_fn(x0, w, b, mul0)
+            y0_expected = fn(x0, w, b, mul0)
+            torch.testing.assert_close(y0, y0_expected)
+
+            x1 = torch.rand(K, M1).cuda().half()
+            mul1 = torch.rand(M1, N).cuda().half()
+            y1 = compiled_fn(x1, w, b, mul1)
+            y1_expected = fn(x1, w, b, mul1)
+            torch.testing.assert_close(y1, y1_expected)
```
Comment (jansel) [inline test/inductor/test_max_autotune.py:852]: Can we add an assert that the prologue was actually generated?  Either by inspecting the output code (there are some helpers to get this) or by counting kernels.

File: torch/_inductor/codegen/triton.py [hunk 6]
```diff
@@ -1254,6 +1254,8 @@ def __init__(
         self.iter_vars_count = itertools.count()
         self.inside_reduction = self.numels[-1] != 1
         self.body = IndentedBuffer()
+        self.prologue_A_body = IndentedBuffer()
+        self.prologue_B_body = IndentedBuffer()
```
Comment (jansel) [inline torch/_inductor/codegen/triton.py:1258]: Can this be done without special-casing the fact that there are two input named A and B?

Some templates might take more (or less) args and name them differently.

File: torch/_inductor/kernel/mm.py [hunk 7]
```diff
@@ -76,6 +76,8 @@
             b = tl.load(B, mask=rk[:, None] < k, other=0.)
         if B_PROLOGUE_CAST_TYPE is not None:
             b = b.to(B_PROLOGUE_CAST_TYPE)
+        {{prologue_A("a")}}
+        {{prologue_B("b")}}
```
Comment (jansel) [inline torch/_inductor/kernel/mm.py:80]: Should there be a flag on the template to say this is supported?  Other templates (like conv) won't support this.

File: torch/_inductor/scheduler.py [hunk 8]
```diff
@@ -843,6 +858,7 @@ def __init__(self, scheduler: "Scheduler", snodes: List[SchedulerNode]):
         self.users: List[NodeUser] = []
         self.inverse_users = []
         self.node_users = []
+        self.rename_dict: [str, set()] = dict()
```
Comment (jansel) [inline torch/_inductor/scheduler.py:861]:
```suggestion
        self.rename_dict: Dict[str, Set[...]] = dict()
```

File: test/inductor/test_max_autotune.py [hunk 9]
```diff
@@ -805,6 +805,52 @@ def fn(
             y1_expected = fn(x1, w, b, mul1)
             torch.testing.assert_close(y1, y1_expected)
 
+    def test_triton_template_with_prologues_and_dynamic_shape(self):
```
Comment (jansel) [inline test/inductor/test_max_autotune.py:808]: Can we add another test where this is an additional usage of the prologue?

File: torch/_inductor/scheduler.py [hunk 10]
```diff
@@ -2008,6 +2021,42 @@ def can_fuse(self, node1: BaseSchedulerNode, node2: BaseSchedulerNode):
...
+                return False
+
+            # Check if input size is a power of 2 and more than the smallest block size, which is defined in filtered_configs in mm_common.py
+            node = node1.get_nodes()[0]
+            if isinstance(node.node, ComputedBuffer):
+                for read in node1.read_writes.reads:
+                    if isinstance(read, MemoryDep):
+                        for size in read.size:
+                            if size & (size - 1) != 0:
+                                why("Tensor size is not a power of 2. Zero-padding will incur wrong results in some prologue functions such as sigmoid.")
```
Comment (jansel) [inline torch/_inductor/scheduler.py:2040]: Is power of 2 enough here?  Couldn't some block sizes result in padding for smaller powers of 2. ?

Perhaps we can codegen using `ops.masked(cond, lambda:..., 0.0)` to zero out th padding after the prologue is run.

**Review by jansel — CHANGES_REQUESTED** [review]
This is neat thanks for the PR!

Fix lints reported by CI.


Do you have any performance results?  I'd expect this to sometimes result in slowdowns since the prologues will be executed many more times than they would as a separate kernel.  We might need some heuristic to decide only enable it for cheap ops.

You can run the benchmarks here with and without this PR (in max-autotune mode):
https://github.com/pytorch/pytorch/tree/main/benchmarks/dynamo

File: torch/_inductor/codegen/triton.py (hunk 4 above)
Comment (michihirohorie) [reply torch/_inductor/codegen/triton.py]: Fixed.

File: torch/_inductor/kernel/mm.py (hunk 7 above)
Comment (michihirohorie) [reply torch/_inductor/kernel/mm.py]: Yes, it is better to have a flag. Fixed by inserting a guard.

File: torch/_inductor/scheduler.py (hunk 8 above)
Comment (michihirohorie) [reply torch/_inductor/scheduler.py]: Thank you! Fixed.

**Comment by jansel** [comment]
Thanks, I'm curious to see the perf effect of this.

File: test/inductor/test_max_autotune.py (hunk 5 above)
Comment (michihirohorie) [reply test/inductor/test_max_autotune.py:850]: I added an assert to check if the fusion count is expected.

File: test/inductor/test_max_autotune.py (hunk 9 above)
Comment (michihirohorie) [reply test/inductor/test_max_autotune.py:805]: I actually updated this test so that it would be more aligned with our target.

File: torch/_inductor/scheduler.py (hunk 10 above)
Comment (michihirohorie) [reply torch/_inductor/scheduler.py]: I updated to use ops.masked() to zero out the padding. Thanks.

**Comment by michihirohorie** [comment]
I took some time to get better understanding how Inductor works. 

I tried pytorch/benchmark/dynamo with the latest changes, but unfortunately did not see performance improvement. It seems to be that prologue fusion is not applied due to limited feature of current PR. 

What we finally want to realize is fusing dequantization and matmul, which is shown in L11-14 of Algorithm 1  in this paper: https://arxiv.org/pdf/2402.00025.pdf. As a first step, I’m trying to enable simple prologue fusion.

**Comment by jansel** [comment]
Are you benchmarking with max-autotune enabled?  If not Triton matmuls don't get used.

**Comment by michihirohorie** [comment]
I set TORCHINDUCTOR_MAX_AUTOTUNE=1, but I'll check what's happening in detail with TORCH_LOGS.
